distributed/diagnostics/plugin.py

```diff
@@ -358,20 +361,106 @@ class PackageInstall(WorkerPlugin, abc.ABC):
     PipInstall
     """
 
-    INSTALLER: ClassVar[str]
+    WORKER_PLUGIN_CLASS: ClassVar[type]
 
+    installer: _PackageInstaller
+    name: str
+    restart_workers: bool
+
+    def __init__(
+        self,
+        installer: _PackageInstaller,
+        restart_workers: bool,
+    ):
+        self.installer = installer
+        self.restart_workers = restart_workers
+        self.name = f"{self.installer.INSTALLER}-install-{uuid.uuid4()}"
+
+    async def start(self, scheduler: Scheduler) -> None:
+        if not hasattr(scheduler, "_package_install_lock"):
+            scheduler._package_install_lock = asyncio.Lock()  # type: ignore[attr-defined]
+
+        async with scheduler._package_install_lock:  # type: ignore[attr-defined]
+            if not self._is_installed(scheduler):
+                logger.info(
+                    "%s installing the following packages on the scheduler: %s",
+                    self.installer.INSTALLER,
+                    self.installer.packages,
+                )
+                self._set_installed(scheduler)
+                self.installer.install()
+            else:
+                logger.info(
+                    "The following packages have already been installed on the scheduler: %s",
+                    self.installer.packages,
+                )
+
+            worker_plugin = _PackageInstallWorker(self.installer, self.restart_workers)
+            await scheduler.register_worker_plugin(None, dumps(worker_plugin))
+
+    def _is_installed(self, scheduler):
+        return scheduler.get_metadata(self._compose_installed_key(), default=False)
+
+    def _set_installed(self, scheduler):
+        scheduler.set_metadata(
+            self._compose_installed_key(),
+            True,
+        )
+
+    def _compose_installed_key(self):
+        return [
+            self.name,
+            "installed",
+            socket.gethostname(),
+        ]
+
+    def setup(self, worker):
+        raise TypeError(
+            "PackageInstall is a scheduler plugin, not a worker plugin; "
+            "use `client.register_scheduler_plugin(...) to register it."
+        )
```

Comment:
This will raise if the plugin gets registered as a worker plugin while workers are alive, but it will get silenced if it gets installed on newly created workers after registration.

Reply:
This bit is superseded by #8169, which will register the plugin as a scheduler plugin.